x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/logging/Throttler.java

```diff
@@ -9,150 +9,188 @@
-        logExecutor.log(message);
+        logExecutor.logFirstMessage();
     }
 
     @Override
     public void close() {
         isRunning.set(false);
-        cancellableTask.get().cancel();
+        if (cancellableTask.get() != null) {
+            cancellableTask.get().cancel();
+        }
         logExecutors.clear();
     }
 
-    private static class LogExecutor {
-        private final long skippedLogCalls;
-        private final Instant timeOfLastLogCall;
+    // default for testing
+    static class LogExecutor {
+        // -1 here because we need to determine if we haven't logged the first time
+        // After the first time we'll set it to 0, then the thread that runs on an interval
+        // needs to know if there are any repeated message, if it sees 0, it knows there are none
+        // and skips emitting the message again.
+        private static final long INITIAL_LOG_COUNTER_VALUE = -1;
+
+        private final AtomicLong skippedLogCalls;
+        private Instant timeOfLastLogCall;
         private final Clock clock;
-        private final Consumer<String> consumer;
+        private final Consumer<String> throttledConsumer;
+        private final String originalMessage;
 
-        LogExecutor(Clock clock, Consumer<String> throttledConsumer) {
-            this(clock, 0, throttledConsumer);
+        LogExecutor(Clock clock, Consumer<String> throttledConsumer, String originalMessage) {
+            this(clock, INITIAL_LOG_COUNTER_VALUE, throttledConsumer, originalMessage);
         }
 
-        LogExecutor(Clock clock, long skippedLogCalls, Consumer<String> consumer) {
-            this.skippedLogCalls = skippedLogCalls;
+        LogExecutor(Clock clock, long skippedLogCalls, Consumer<String> throttledConsumer, String originalMessage) {
+            this.skippedLogCalls = new AtomicLong(skippedLogCalls);
             this.clock = Objects.requireNonNull(clock);
             timeOfLastLogCall = Instant.now(this.clock);
-            this.consumer = Objects.requireNonNull(consumer);
+            this.throttledConsumer = Objects.requireNonNull(throttledConsumer);
+            this.originalMessage = Objects.requireNonNull(originalMessage);
         }
 
-        void log(String message) {
-            this.consumer.accept(message);
+        void logRepeatedMessages() {
+            long numSkippedLogCalls;
+            synchronized (skippedLogCalls) {
+                numSkippedLogCalls = skippedLogCalls.get();
+                if (hasRepeatedLogsToEmit(numSkippedLogCalls) == false) {
+                    // Since we tried to log but there were no repeated messages we'll reset this entry so a new message
+                    // would get logged like it is the first time
+                    skippedLogCalls.set(INITIAL_LOG_COUNTER_VALUE);
+                    return;
+                }
+            }
+
```

Comment:
If a message is about to check the first time logic it will get skipped here because we aren't reseting the log counter. So technically we'd miss one repeat count but that seems fine.